Rewrite to use `remark-language-server` (#65)
remarkjs/vscode-remark: readme.md

> # Remark for Visual Studio Code

Reviewer: ```suggestion
# vscode-remark
```

Should be package/extension name?

Author: I looked at what other extensions use, mainly the ones published by well known publishers and official extensions. `{tool} for Visual Studio Code` or something similar appears to be the most used level 1 heading in their readmes.

Reviewer: Looked through the top 2 pages of the microsoft org on GH, by stars, searching for `vscode`.
I also find that there’s no consistency. A lot of random one-off weird things, but here’s the ones I mostly found:

1. `remark for Visual Studio Code` (https://github.com/microsoft/vscode-mssql#mssql-for-visual-studio-code, https://github.com/microsoft/vscode-docker#docker-for-visual-studio-code----, https://github.com/microsoft/vscode-postgresql#postgresql-for-visual-studio-code, https://github.com/microsoft/vscode-tools-for-ai#azure-machine-learning-for-visual-studio-code, https://github.com/microsoft/vscode-azurefunctions#azure-functions-for-visual-studio-code)
2. `remark for VS Code` (https://github.com/microsoft/vscode-sublime-keybindings, https://github.com/microsoft/vscode-cosmosdb, https://github.com/microsoft/vscode-atom-keybindings)
3. `vscode-remark` (https://github.com/microsoft/vscode-css-languageservice#vscode-css-languageservice, https://github.com/microsoft/vscode-json-languageservice#vscode-json-languageservice)

I’m open to 1. Though, as 3. is used sometimes by them and everywhere by us, that has my preference.

Author: I used 1, as the references for option 3 don’t link to VSCode plugins.

[![Build][build-badge]][build]
[![Coverage][coverage-badge]][coverage]
[![Downloads][downloads-badge]][downloads]
[![Sponsors][sponsors-badge]][collective]
[![Backers][backers-badge]][collective]
[![Chat][chat-badge]][chat]

**[remark][]** is a tool that transforms markdown with plugins.
These plugins can inspect and change your markup.
You can use remark on the server, the client, CLIs, deno, etc.

> This plugin for Visual Studio Code lets you format and lint markdown files using
> remark.

Reviewer: ```suggestion
This plugin for Visual Studio Code lets you format and lint markdown files with
remark.
```

Reviewer: This whole paragraph could be shorter, potentially. Just focussing on this exact project, rather than remark.

Reviewer: a “what is this?” section can go more in-depth

## Contents

*   [When should I use this?](#when-should-i-use-this)
*   [Use](#use)
    *   [Formatting](#formatting)
*   [Plugins](#plugins)
*   [Syntax](#syntax)
*   [Compatibility](#compatibility)
*   [Security](#security)
*   [Contribute](#contribute)
*   [License](#license)

## When should I use this?

If your project is checked using [`remark-cli`][remark-cli] and you use Visual

> Studio Code to edit markdown files, it’s recommended to use this extension.

Reviewer: Maybe folks don’t know about `remark-cli`.
Maybe this can be reworded without assuming that knowledge, to explain why checking/formatting markdown is useful

Author: I reworded this. Please re-review. :)

## Use

To start using this for a project, [`remark-cli`][remark-cli] needs to be
configured first.

> For example, install `remark-cli` and `remark-preset-lint-recommended`:

Reviewer: It’s not clear why `lint-recommended` is needed. As this example isn’t just about VS Code, it might be good to explain more what this example shows

```sh
npm install remark-cli remark-preset-lint-recommended
```

> Now create a file named `.remarkrc.json` with the following content:

Reviewer: While `.json` isn’t bad, it feels more of a personal convention to use it here. This VS Code extension highlights `.remarkrc` as well. Should it be explained why this file is named that way?

Author: I had to pick something. I think `.remarkrc.json` is understood by most people.

I don’t think we should suggest `.remarkrc` just because this extension supports it. I think it’s best to always use an extension. `.eslintrc` without extension has even been deprecated.

I don’t think we need to explain why the file is named that way. It’s a common convention.

Reviewer: > I don’t think we need to explain why the file is named that way. It’s a common convention.

I think people will come into this extension not knowing what remark is. Or knowing that RC files exist.

```json
{

>   "plugins": [["preset-lint-recommended"]]

Reviewer: extra brackets are not needed if there’s no config.

}
```

Now any syntax which doesn’t comply with `remark-preset-lint-recommended` will

> be reported in the *Problems* pane in your editor.

Reviewer: Should this explain that a user should open a readme or whatever? Perhaps good to start with a markdown example, and then here show what the output in `Problems` is, if everything is working perfectly

### Formatting

This extension can be used as a formatter for markdown files.

> To format a file, first press <kbd>Ctrl</kbd>+<kbd>Shift</kbd>+<kbd>P</kbd>.

Reviewer: is formatted a bit verbose, why not one `kbd` with the plusses inside?

Author: I didn’t even know if your suggestion is valid, so I googled `html kbd`. It turnes out some people use it, but more people favor the format I’m using here.

It also makes more sense to me. You don’t press a single key which is labeled <kbd>Ctrl + Shift + P</kbd>. You press 3 keyboard keys: <kbd>Ctrl</kbd>+<kbd>Shift</kbd>+<kbd>P</kbd> (maybe some spacing would be nice though)

Reviewer: Interesting!!

> ou don’t press a single key which is labeled Ctrl + Shift + P. You press 3 keyboard keys: Ctrl+Shift+P (maybe some spacing would be nice though)

But you have to press them at the same time. Your latter example to me indicated something done in sequence rather than at the same time.

>  but more people favor the format I’m using here.

Could you enlighten me on what you found, that indicates this?

---

The first example two examples in the HTML spec is way more complex than what we both do: https://html.spec.whatwg.org/multipage/text-level-semantics.html#the-kbd-element.
The third example, seems to be more in line with what I’d suggest. Prefixed with “Such precision isn't necessary; the following is equally fine”

Reviewer: Interestingly the readme of this project currently uses my style: https://github.com/remarkjs/vscode-remark/blob/4830030f626cc24bff3eb23f8dd99c42fa69af09/README.md?plain=1#L20-L21.
I do find a lot of examples of your style though.

Reviewer: One more reason in favor of my style: the plusses are weird in your style. They’re not natural language. Why not use `and`? Or commas? Makes more sense to me when they’re inside the `kbd`

Author: Some examples which use my style are the [kbd docs on mdn](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/kbd), [kbd docs on w3schools](https://www.w3schools.com/tags/tag_kbd.asp) and [Wikipedia article on Ctrl-Alt-Delete](https://en.wikipedia.org/wiki/Control-Alt-Delete). IMHO this also makes sense, as it’s rendered to look similar to a key on a keyboard.

I’m also pretty sure this is the style I see most of the time, although it’s hard to find examples when actually looking for them.

From https://html.spec.whatwg.org/multipage/text-level-semantics.html#the-kbd-element:

> When the `kbd` element is nested inside another `kbd` element, it represents an actual key or other single unit of input as appropriate for the input mechanism.
> […]
> > <p>To make George eat an apple, press <kbd><kbd>Shift</kbd>+<kbd>F3</kbd></kbd></p>

So then I’d say it should be:

<kbd><kbd>Ctrl</kbd>+<kbd>Shift</kbd>+<kbd>P</kbd></kbd>

This also looks fine with me.

Reviewer: Yeah that last one looks a bit verbose to me (which the intro for the 3rd example also seems to implicated). I'm fine with "your style" 👍

Search for and select *“Format Document With…”*.

> Now select *“Remark”*.

Reviewer: Menu items could also use `` ` `` (code)

To make this the default formatter, add the following to your `settings.json`

> (Open with <kbd>Ctrl</kbd>+<kbd>,</kbd>):

Reviewer: same as above

```json
{
  "[markdown]": {
    "editor.defaultFormatter": "unifiedjs.vscode-remark",
  }
}
```

Optionally enable format on save:

```json
{
  "[markdown]": {
    "editor.defaultFormatter": "unifiedjs.vscode-remark",
    "editor.formatOnSave": true
  }
}
```

## Plugins

The **remark** ecosystem has a variety of plugins available.
Most notably you’ll want to check out [`remark-lint`][remark-lint].

> A curated list of plugins can be found [here][list-of-plugins].

Reviewer: “here” as link text is confusing. The “List of plugins” should be linked instead. Perhaps reword as: `See the [List of plugins][...] for all remark plugins`

## Syntax

remark follows CommonMark, which standardizes the differences between markdown
implementations, by default.
Some syntax extensions are supported through plugins.

## Compatibility

Projects maintained by the unified collective are compatible with all maintained
versions of Node.js.

> As of now, that is Node.js 12.20+, 14.14+, and 16.0+.

Reviewer: Is this compat range useful for VS Code users? Should this be about VS Code support?

Author: I’m not entirely sure. `remark-language-server` is bundled with the extension. The NodeJS support ranges do apply to this.

Also I don’t really know how VSCode handles compatibility for extensions.

Reviewer: VS Code does bundle Node though, so Node.js doesn‘t matter, but VS Code should?
Also: isn’t esbuild used?

Author: The `node` executable is used to run the language server. I don’t think VS Code injects its own `PATH`.

That’s what’s nice about using a language server: it doesn’t depend on the editor. Both happen to use Node.js, but that’s a coincidence.

I wouldn’t rely on esbuild for compatibility. This can likely fix syntactic compatibility issues, but not for example new methods introduced in Node.js 12.20+, which unified may depend on.

---

Looking at the answer I just typed I think the answer is simply: Yes, the Node.js compatibility range is useful.

Reviewer: OK, sounds good.
I feel like it would make sense to include a VSC version of it applies and is findable someway

Author: I documented it works with VS Code 1.46.0 and greater, because that’s what’s documented in the `package.json` engines field. I don’t know what this means in practice. I think most people update VS Code regularly anyway. :shrug:

Reviewer: Hmm, so the `engines` field is likely wrong as well? :(

Author: It’s probably fine. I’m guessing bootstrapping a new project simply sets it to the current latest version of VS Code.

# Remark for Visual Studio Code

[![Build][build-badge]][build]
[![Coverage][coverage-badge]][coverage]
[![Downloads][downloads-badge]][downloads]
[![Sponsors][sponsors-badge]][collective]
[![Backers][backers-badge]][collective]
[![Chat][chat-badge]][chat]

This plugin for Visual Studio Code lets you format and lint markdown files with
remark.

## Contents

*   [What is this?](#what-is-this)
*   [When should I use this?](#when-should-i-use-this)
*   [Use](#use)
    *   [Formatting](#formatting)
*   [Plugins](#plugins)
*   [Syntax](#syntax)
*   [Compatibility](#compatibility)
*   [Security](#security)
*   [Contribute](#contribute)
*   [License](#license)

## What is this?

**[remark][]** is a tool that transforms markdown with plugins.
These plugins can inspect and change your markup.

## When should I use this?

> If you wish to apply code quality checks on your markdown files, this extension
> is for you.
> This extension allows you to check and format markdown files right in your
> editor.
> [`remark-cli`][remark-cli] can be used to ensure consitent formatting in CI.
> it’s recommended to use that in combination with this extension.

Reviewer: How about:

```suggestion
You can use this extension if you want to check (lint) and format markdown
files from within your editor.

To configure this extension, you define your preferred markdown style in a
configuration file (`.remarkrc`, `.remarkrc.js`, etc. or in `package.json`).
This file is picked up by `vscode-remark` and other tools (useful for
contributors that don’t use VS Code).

The configuration file is also used by [`remark-cli`][remark-cli], which is
recommended to be used alongside `vscode-remark`, as an npm script and/or in
CI, to enforce the markdown style.
```

- More concrete language in the first two sentences
- More thorough explanation of when to use this extension / what else to use, too

*   [Use](#use)
    *   [Formatting](#formatting)
*   [Plugins](#plugins)
*   [Syntax](#syntax)
*   [Compatibility](#compatibility)
*   [Security](#security)
*   [Contribute](#contribute)
*   [License](#license)

## What is this?

This project is a Visual Studio Code (VS Code) extension that you can use in
your editor to inspect and change markdown files.
This extension is built around remark, which is a very popular ecosystem of
plugins that work with markdown.
You can choose from the 150+ plugins that already exist or make your own.

See [remark][] for info on what the remark ecosystem is.

## When should I use this?

You can use this extension if you want to check (lint) and format markdown
files from within your editor.

To configure this extension, you define your preferred markdown style in a
configuration file (`.remarkrc`, `.remarkrc.js`, etc. or in `package.json`).
This file is picked up by `vscode-remark` and other tools (useful for
contributors that don’t use VS Code).

The configuration file is also used by [`remark-cli`][remark-cli], which is
recommended to be used alongside `vscode-remark`, as an npm script and/or in
CI, to enforce the markdown style.

> 

Reviewer: Should be an install section here? Perhaps VSC comes with a CLI as well, to do that with a command, and/or perhaps an example of how to add it with the GUI?

Author: I thought I added one already, but I was wrong :thinking:. Anyway, there are multiple ways to get there. I added one method, as I don’t think it’s necessary to document all methods a VS Code extension can be installed.

Also the method I chose works for open source VS Code forks (That’s another issue I’ll create later).

Reviewer: Linking to it might be useful as well? `vscode:extension/xxxx`? Or linking to the marketplace?:

````
[Get it from the VS Code Marketplace](https://marketplace.visualstudio.com/items?itemName=xxxx)

Or launch Quick Open (<kbd>Ctrl</kbd> + <kbd>P</kbd>) and run:

```sh
ext install unifiedjs.vscode-remark
```
````

Author: I didn’t know about that link format. It appears to work in the VS Code markdown preview. However, GitHub doesn’t render it.

I now modelled it after https://github.com/prettier/prettier-vscode#installation

## Use

> To start using this for a project, [`remark-cli`][remark-cli] needs to be
> configured first.
> First, install `remark-cli` and a plugin or preset.
> Let’s use `remark-preset-lint-recommended` for our example.
> 
> ```sh
> npm install remark-cli remark-preset-lint-recommended
> ```
> 
> Now create a file named `.remarkrc.json` with the following content:
> 
> ```json
> {
>   "plugins": ["preset-lint-recommended"]
> }
> ```
> 
> Now let’s create a markdown file `readme.md` and add the following content:
> 
> ```markdown
> # My project
> 
> This is a [broken link][]
> 
> - A `*` should be used for list items
> - Also 4 spaces should be used for indentation
> 
> [unused reference]: https://example.com
> ```
> 
> Now any syntax which doesn’t comply with `remark-preset-lint-recommended` will
> be reported using squiggly lines in your editor and in the `Problems` pane.

Reviewer: ````suggestion
This example enables VS Code users that have this extension installed to see
warnings about markdown style in their editor and the `Problems` pane and to
format markdown from a command.

Make sure that this extension is installed and then install the CLI and
plugins:

```sh
npm install remark-cli remark-preset-lint-recommended remark-preset-lint-consistent remark-toc
```

Then, add a `remarkConfig` to your `package.json` to configure remark:

```js
  /* … */
  "remarkConfig": {
    "settings": {
      "bullet": "*", // Use `*` for list item bullets (default)
      // See <https://github.com/remarkjs/remark/tree/main/packages/remark-stringify> for more options.
    },
    "plugins": [
      "remark-preset-lint-consistent", // Check that markdown is consistent.
      "remark-preset-lint-recommended", // Few recommended rules.
      [
        // Generate a table of contents in `## Contents`
        "remark-toc",
        {
          "heading": "contents"
        }
      ]
    ]
  },
  /* … */
```

> 👉 **Note**: you must remove the comments in the above examples when
> copy/pasting them, as comments are not supported in `package.json` files.

Now, you can open markdown files in your project, and you’ll see squiggly lines
and warnings in the `Problems` pane.
Here’s an example that should produce problems you can use to verify:

```markdown
1) Hello, _Jupiter_ and *Neptune*!
```
````

Reviewer: - This is based on the main example in `remark-cli` and some examples in `remark-lint`.
- As we assume we’re in an npm package, we can place config in `package.json` instead of a separate file
- This shows more examples of plugins/presets, including how to pass options
- Q: is `remark-cli` needed here?

Author: `remark-parse` and `remark-stringify` are needed, which are pulled in by `remark-cli`. I’d just suggest to install `remark-cli`.

> ### Formatting

Reviewer: Not sure this should be an H3 under `## Use` 🤔. Maybe an H2?

Author: It should definitely under `Use`, as it’s a way to use the extension. I added a H3 `Linting`, because the section on top of this explains the linting part.

Reviewer: `Use` should be a short but useful explanation of how to minimally use a thing.
It’s not a place to list every possible way something could be used.

You can instead use `API`, `CLI`, etc. if you want to expand on the interface in detail.
If this is about examples, use an `## Examples`, with another `### Example: formatting ...`

Author: IMHO it feels like you’re trying to force this into a format that doesn’t really work. Both **linting** and **formatting** are **use** cases of this extension.

Anyway, I don’t feel like this is worth the discussion, so I removed the `Linting` header and made the `Formatting` header level 2.

> This extension can be used as a formatter for markdown files.

Reviewer: ```suggestion
This extension can format markdown files.
```

> To format a file, first press <kbd>Ctrl</kbd> + <kbd>Shift</kbd> + <kbd>P</kbd>.

Reviewer: ```suggestion
To format a file, first press <kbd>Ctrl</kbd> + <kbd>Shift</kbd> + <kbd>p</kbd>.
```
```suggestion
To format a file, first press <kbd>Ctrl</kbd> + <kbd>P</kbd>.
```

`shift + p` is implied with `P`?

Reviewer 2: I think it's more user-friendly to be verbose and list all keys to be pressed. Whether the `P` should be lowercase or uppercase is a wish wash; the letter on the physical key is usually an uppercase `P`, but pressing it without shift or caps lock outputs a lowercase `p`. I think I prefer to label what's written on the physical key, but know that's not a water-tight strategy for all symbols across all keyboard layouts.

Author: It’s a keyboard shortcut, not an uppercase `P` character somehow combined with the <kbd>Ctrl</kbd> key. Most of the time alphabetic keyboard keys are indicated as upper case, both on the keyboard and in documentation.

> To format a file, first press <kbd>Ctrl</kbd> + <kbd>Shift</kbd> + <kbd>P</kbd>.

Reviewer: Also:

```
To format a file, open it, press …
```

Search for and select `Format Document With…`.
Now select `remark`.

> To make this the default formatter, add the following to your `settings.json`
> (Open with <kbd>Ctrl</kbd> + <kbd>,</kbd>):

Reviewer: ```suggestion
To make `vscode-remark` the default formatter for markdown, add the following to
your `settings.json` (which you can open with <kbd>Ctrl</kbd> + <kbd>,</kbd>):
```

```json
{
  "[markdown]": {
    "editor.defaultFormatter": "unifiedjs.vscode-remark",
  }
}
```

> Optionally enable format on save:

Reviewer: ```suggestion
You can optionally choose to automatically format when saving with
`editor.formatOnSave`:
```

Reviewer: Are there other options? Should there be an API section? Something else?

Author: Yes, there are also `editor.formatOnPaste` and `editor.formatOnType`. I personally don’t use them, and I’ve seen there are special LSP calls for `onTypeFormatting`, so I don’t even know if that works. I wouldn’t recommend either anyway.

*   [Compatibility](#compatibility)
*   [Security](#security)
*   [Contribute](#contribute)
*   [License](#license)

## What is this?

This project is a Visual Studio Code (VS Code) extension that you can use in
your editor to inspect and change markdown files.
This extension is built around remark, which is a very popular ecosystem of
plugins that work with markdown.
You can choose from the 150+ plugins that already exist or make your own.

See [remark][] for info on what the remark ecosystem is.

## When should I use this?

You can use this extension if you want to check (lint) and format markdown
files from within your editor.

To configure this extension, you define your preferred markdown style in a
configuration file (`.remarkrc`, `.remarkrc.js`, etc. or in `package.json`).
This file is picked up by `vscode-remark` and other tools (useful for
contributors that don’t use VS Code).

The configuration file is also used by [`remark-cli`][remark-cli], which is
recommended to be used alongside `vscode-remark`, as an npm script and/or in
CI, to enforce the markdown style.

## Install

[Get it on the VS Code Marketplace][marketplace] or install it by using Quick

> Open (<kbd>Ctrl</kbd> + <kbd>P</kbd>) and running the following:
> ```txt

Reviewer: ````suggestion
Open (<kbd>Ctrl</kbd> + <kbd>P</kbd>) and running the following:

```txt
````

```markdown
1) Hello, _Jupiter_ and *Neptune*!
```

## Formatting

This extension can format markdown files.

To format a file, pull up the command pallete (<kbd>Ctrl</kbd> +
<kbd>Shift</kbd> + <kbd>P</kbd>), choose `Format Document With…`, and select
`remark`.

To make `vscode-remark` the default formatter for markdown, add the following to
your `settings.json` (which you can open with <kbd>Ctrl</kbd> + <kbd>,</kbd>):

```json
{
  "[markdown]": {
    "editor.defaultFormatter": "unifiedjs.vscode-remark",
  }
}
```

Now markdown documents can be formatted using <kbd>Ctrl</kbd> +
<kbd>Shift</kbd> + <kbd>I</kbd>.

You can optionally choose to automatically format when saving with
`editor.formatOnSave`:

```json
{
  "[markdown]": {
    "editor.defaultFormatter": "unifiedjs.vscode-remark",
    "editor.formatOnSave": true
  }
}
```

## Plugins

The **remark** ecosystem has a variety of plugins available.
Most notably you’ll want to check out [`remark-lint`][remark-lint].
See this curated [list of plugins][list-of-plugins] for more remark plugins.

## Syntax

remark follows CommonMark, which standardizes the differences between markdown
implementations, by default.

> Some syntax extensions are supported through plugins.

Reviewer: I’m not too sure how much this contributes here, I’m open to removing it, but also okay to keep it.

Author: Do you mean the entire `Syntax` section? To me it doesn’t really make a difference. It’s documented in other places as well.

Maybe it’s best to remove it to keep it simple?

Reviewer: yep, I mean the whole section. So sounds like we agree on removing it?

Author: Yep, removed.

You can choose from the 150+ plugins that already exist or make your own.

See [remark][] for info on what the remark ecosystem is.

## When should I use this?

You can use this extension if you want to check (lint) and format markdown
files from within your editor.

To configure this extension, you define your preferred markdown style in a
configuration file (`.remarkrc`, `.remarkrc.js`, etc. or in `package.json`).
This file is picked up by `vscode-remark` and other tools (useful for
contributors that don’t use VS Code).

The configuration file is also used by [`remark-cli`][remark-cli], which is
recommended to be used alongside `vscode-remark`, as an npm script and/or in
CI, to enforce the markdown style.

## Install

[Get it from the VS Code Marketplace](https://marketplace.visualstudio.com/items?itemName=xxxx).
Install through VS Code extensions.
Search for `remark`.

It can also be installed in VS Code: Launch VS Code Quick Open
(<kbd>Ctrl</kbd> + <kbd>P</kbd>), paste the following command, and press enter.

```txt
ext install unifiedjs.vscode-remark
```

See the
[extension installation guide](https://code.visualstudio.com/docs/editor/extension-gallery)

> for details.

Author: Done

## Use

> This example enables VS Code users that have this extension installed to see

Author: Yes, I like that. I did keep the markdown example, because it’s very likely to actually show problems, whereas opening a file in a project where `remark-cli` is configured probably adheres to those rules.

plugins:

```sh
npm install remark-cli remark-preset-lint-recommended remark-preset-lint-consistent remark-toc
```

Then, add a `remarkConfig` to your `package.json` to configure remark:

```js
  /* … */
  "remarkConfig": {
    "settings": {
      "bullet": "*", // Use `*` for list item bullets (default)
      // See <https://github.com/remarkjs/remark/tree/main/packages/remark-stringify> for more options.
    },
    "plugins": [
      "remark-preset-lint-consistent", // Check that markdown is consistent.
      "remark-preset-lint-recommended", // Few recommended rules.
      [
        // Generate a table of contents in `## Contents`
        "remark-toc",
        {
          "heading": "contents"
        }
      ]
    ]
  },
  /* … */
```

> 👉 **Note**: you must remove the comments in the above examples when
> copy/pasting them, as comments are not supported in `package.json` files.

Now, you can open markdown files in your project, and you’ll see squiggly lines
and warnings in the `Problems` pane.
Here’s an example that should produce problems you can use to verify:

```markdown
1) Hello, _Jupiter_ and *Neptune*!
```

## Formatting

This extension can format markdown files.

> To format a file, first press <kbd>Ctrl</kbd> + <kbd>Shift</kbd> + <kbd>P</kbd>.
> Search for and select `Format Document With…`.
> Now select `remark`.

Reviewer: ```suggestion
To format a file, pull up the command pallete (<kbd>Ctrl</kbd> +
<kbd>Shift</kbd> + <kbd>P</kbd>) and choose: `remark: beautify markdown code`
```

Why not recommend one step less?
Inspired by the wording used in https://marketplace.visualstudio.com/items?itemName=samverschueren.linter-xo.
Note: this command (`Remark: Beautify markdown code`) is based on the old extension, what is it now? I don’t really like the wording of `Beautify` and feel like `code` is superfluous?

Author: The instructions you’re proposing are simply wrong. They describe the old behavior of the extension, which is to register the command: `Remark: Beautify markdown code`. The new behavior is to register a formatter, which can be used as is currently described or can be set as the default formatter as described below.

Reviewer: Why is the command removed?

---

Even if now 3 steps are required, the wording from `linter-xo` can be used as inspiration:

```suggestion
To format a file, pull up the command pallete (<kbd>Ctrl</kbd> +
<kbd>Shift</kbd> + <kbd>P</kbd>), choose `Format Document With…`, and select
`remark`.
```

Author: > Why is the command removed?

Because it has been replaced by a formatter.

To make `vscode-remark` the default formatter for markdown, add the following to

> your `settings.json` (which you can open with <kbd>Ctrl</kbd> + <kbd>,</kbd>):

Author: No. Going even more in depth is more something to explain in a VS Code guide. These are the kind of things that are built into the editor and people just know.

> your `settings.json` (which you can open with <kbd>Ctrl</kbd> + <kbd>,</kbd>):

Reviewer: I don’t get why you explain complex key combos in many places while not wanting to explain this.

> Going even more in depth is more something to explain in a VS Code guide.

Then link to it.

> These are the kind of things that are built into the editor

Everything is build into the editor.

> people just know.

Please enlighten me, because I don’t know.

> your `settings.json` (which you can open with <kbd>Ctrl</kbd> + <kbd>,</kbd>):

Author: I added it anyway, although I feel this is getting really verbose.
